File: HEAP (PRIORITY QUEUE)/min_heap.py
class MinHeap:
    def __init__(self,n):
        self.size = 0
        self.totol_size = n
        self.arr = []

    def heapify(self,index):
        smallest = index
        left = 2*index + 1
        right = 2*index + 2
        
        if(left < self.size and self.arr[left] < smallest):
            smallest = left
        if(right < self.size and self.arr[right] < smallest):
            smallest = right
        
        if(smallest == index):
            return
        self.arr[index],self.arr[smallest]=self.arr[smallest],self.arr[index]
        self.heapify(smallest)

    def insert(self,value):
        if(self.size == self.totol_size):
            print("Heap Overflow\n")
            return
        self.arr.append(value)
        index = self.size
        self.size+=1

        # compare it with its parent
        while index > 0 and self.arr[(index-1)//2] > value:
            self.arr[index],self.arr[(index-1)//2]=self.arr[(index-1)//2],self.arr[index]
            index = (index-1)//2
        print(f"{value} inserted into heap")
    def show(self):
        for item in self.arr:
            print(item, end=" ")
        
    def delete(self):
        if(self.size==0):
            print("Heap Underflow")
            return
        self.arr[0]=self.arr[-1]
        self.arr.pop()
        self.size-=1

        if(self.size==0):
            return
        
        self.heapify(0)

h = MinHeap(5)
h.insert(90)
h.insert(100)
h.insert(40)
h.insert(20)
h.show()
        

File: HEAP (PRIORITY QUEUE)/min_heap.cpp
#include <iostream>
using namespace std;
class MinHeap
{
    int *arr;
    int size;
    int capacity;

public:
    MinHeap(int n)
    {
        arr = new int[n];
        size = 0;
        capacity = n;
    }
    void heapify(int index)
    {
        int smallest = index;
        int left = 2 * index + 1;
        int right = 2 * index + 2;

        if (left < size && arr[left] < arr[smallest])
        {
            smallest = left;
        }
        if (right < size && arr[right] < arr[smallest])
        {
            smallest = right;
        }

        if (smallest == index)
        {
            return;
        }
        swap(arr[index], arr[smallest]);
        heapify(smallest);
    }
    void insert(int value)
    {
        if (size == capacity)
        {
            cout << "Heap Overflow" << endl;
            return;
        }
        arr[size] = value;
        int index = size;
        size++;

        while (index > 0 && arr[(index - 1) / 2] > value)
        {
            swap(arr[index], arr[(index - 1) / 2]);
            index = (index - 1) / 2;
        }
        cout << value << " inserted into heap" << endl;
    }
    void print()
    {
        for (int i = 0; i < size; i++)
            cout << arr[i] << " ";
        cout << endl;
    }
    void remove()
    {
        if (size == 0)
        {
            cout << "Heap Underflow..." << endl;
            return;
        }
        arr[0]=arr[size-1];
        size--;
        if(size==0){
            return;
        }

        heapify(0);
    }
};

int main()
{
    MinHeap h(5);
    h.insert(90);
    h.insert(100);
    h.insert(40);
    h.insert(20);
    h.print();

    return 0;
}